neo/Network/LocalNode.cs

```diff
@@ -116,49 +116,89 @@ public LocalNode()
         private static bool AddTransaction(Transaction tx)
         {
             if (Blockchain.Default == null) return false;
-            lock (mem_pool)
+            lock (Blockchain.Default.PersistLock)
             {
-                if (mem_pool.ContainsKey(tx.Hash)) return false;
-                if (Blockchain.Default.ContainsTransaction(tx.Hash)) return false;
-                if (!tx.Verify(mem_pool.Values)) return false;
-                mem_pool.Add(tx.Hash, tx);
-                CheckMemPool();
+                lock (mem_pool)
+                {
+                    if (mem_pool.ContainsKey(tx.Hash)) return false;
+                    if (Blockchain.Default.ContainsTransaction(tx.Hash)) return false;
+                    if (!tx.Verify(mem_pool.Values)) return false;
+                        mem_pool.Add(tx.Hash, tx);
+                    CheckMemPool();
+                }
             }
             return true;
         }
 
         private void AddTransactionLoop()
         {
+            int lastTransactionCount = 0;
+            uint lastBlockHeight = 0;
+            
             while (!cancellationTokenSource.IsCancellationRequested)
             {
                 new_tx_event.WaitOne();
-                Transaction[] transactions;
-                lock (temp_pool)
-                {
-                    if (temp_pool.Count == 0) continue;
-                    transactions = temp_pool.ToArray();
-                    temp_pool.Clear();
-                }
+
                 ConcurrentBag<Transaction> verified = new ConcurrentBag<Transaction>();
-                lock (mem_pool)
+                lock (Blockchain.Default.PersistLock)
                 {
-                    transactions = transactions.Where(p => !mem_pool.ContainsKey(p.Hash) && !Blockchain.Default.ContainsTransaction(p.Hash)).ToArray();
-                    if (transactions.Length == 0) continue;
+                    lock (mem_pool)
+                    {
+                        Transaction[] transactions;
+                        lock (temp_pool)
+                        {
+                            temp_pool.UnionWith(mem_pool.Values);
+                            mem_pool.Clear();
```

Comment:
Every time new transactions arrive, we're going to verify the transactions that have been verified in the mem_pool again?

Reply:
If the transaction verification code of a new transaction ensures that all transactions still in the mem_pool are with absolute certainty not invalidated by the new transaction, then I could adjust this to only verifying the new transactions from the temp_pool. However, I don't think that is the case, any time a new transaction is added all existing transactions have to be re-verified; that is what the code is already doing before this change, notice here:
https://github.com/neo-project/neo/blob/master/neo/Network/LocalNode.cs#L148
Unless I'm missing something, the previous code always did a union with the transactions in the mem_pool and verified them again also...

I noticed this double verification thing myself also, but since the add verification code needs to be able to return an error if the transaction is invalid and it doesn't want to take the time to verify all transactions in the mem_pool I don't see a way around it.

Reply:
No, the code in https://github.com/neo-project/neo/blob/master/neo/Network/LocalNode.cs#L148 just using the mem_pool as the parameter of `Transaction.Verify()`.

The transactions in mem_pool are verified, and do not need to be verified again when new transactions arrive. Because if a new transaction conflicts with the old ones, the new transaction will be rejected. That's why we're going to pass the mem_pool into that function `Transaction.Verify()`.

Reply:
Ah, thank you for that explanation. I will adjust this now and update the pull request.